keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,605 @@
+
+### Unit Tests
+
+This is subject to the new re-architected controller implementation. Overall these scenarios would be tested.
+
+- [Run or Not](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L167) tests the controller under different scenarios to check if the Job is created or not
+- [Validates Job cleanup](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L371) path of the controller under different conditions.
+- [Validates Status](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/pkg/controller/cronjob/cronjob_controller_test.go#L593) of the CronJob after sync under different conditions.
+
+## Implementation plan
+
+#### Release 1.20
+
+- Alpha: Feature flag for new controller is disabled by default
+- Dual controller. Both old and new implementation co-exist
+
+#### Release 1.21
+
+- Beta: Feature flag for new controller is enabled by default. If the distribution chooses it can be disabled
+- Introduce CronJob in batch/v1
+
+#### Release 1.22
+
+- GA: The feature flag is deprecated and the old controller code is cleaned up
+
+## Graduation Criteria
+
+- [ ] Implement shared informers to reduce pressure on API Server
+- [ ] Pass conformance tests
+- [ ] Update documents reflecting the changes
+- [ ] Pass CronJob e2e tests
+- [ ] Pass CronJob unit-tests
+- [ ] Pass scale tests
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: `CronJobControllerV2`
+    - Components depending on the feature gate: `kube-controller-manager`
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  No changes are expected.
+
+* **Are there any tests for feature enablement/disablement?**
+  There are unit tests at minimum that verify full correctness of the new controller. Additionally, we want to ensure the new controller passes all the existing e2e tests for the current controller.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  The only time when this might fail is when leader election for kube-controller-manager ends up assigned to two separate instances. In this case the two controller might end up fighting over the cronjob resource.
+
+* **What specific metrics should inform a rollback?**
+  None.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  I do not believe this is applicable.
```

Comment:
This is more-or-less similar to components restart, but it still would be useful thing to do (at least manually).

Can you change it to "Manual test is planned once the implementation is finished" and add that for Beta graduation criteria?

Reply:
Added, and graduation criteria already cover units, conformance and e2e.